Cython/Utility/FunctionArguments.c

```diff
@@ -323,53 +424,93 @@ static int __Pyx_ParseOptionalKeywords(
 #endif
         Py_INCREF(value);
 
+        if (unlikely(!PyUnicode_Check(key))) goto invalid_keyword_type;
+
+        #if CYTHON_USE_UNICODE_INTERNALS
+        // The key hash is probably pre-calculated.
+        key_hash = ((PyASCIIObject*)key)->hash;
+        if (unlikely(key_hash == -1)) {
+            key_hash = PyObject_Hash(key);
+            if (unlikely(key_hash == -1))
+                goto bad;
+        }
+        #endif
+
+        // Compare strings for non-interned matches.
         name = first_kw_arg;
-        if (likely(PyUnicode_Check(key))) {
-            while (*name) {
-                int cmp = (
-                #if CYTHON_ASSUME_SAFE_SIZE
-                    (PyUnicode_GET_LENGTH(**name) != PyUnicode_GET_LENGTH(key)) ? 1 :
-                #endif
-                    PyUnicode_Compare(**name, key)
-                );
+        while (*name) {
+            PyObject *name_str = **name;
+            #if CYTHON_USE_UNICODE_INTERNALS
+            // Our argument hash is definitely pre-calculated.
+            if (key_hash == ((PyASCIIObject*)name_str)->hash)
+            #elif CYTHON_ASSUME_SAFE_SIZE
+            if (PyUnicode_GET_LENGTH(name_str) == PyUnicode_GET_LENGTH(key))
+            #endif
+            {
+                #if CYTHON_USE_UNICODE_INTERNALS
+                if (__Pyx_UnicodeKeywordsEqual(name_str, key))
+                #else
+                int cmp = PyUnicode_Compare(name_str, key);
                 if (cmp < 0 && unlikely(PyErr_Occurred())) goto bad;
-                if (cmp == 0) {
+                if (cmp == 0)
```

Comment:
Would it more readable to make this stuff the definition of `__Pyx_UnicodeKeywordsEqual` when not `CYTHON_USE_UNICODE_INTERNALS`

Reply:
Currently, `__Pyx_UnicodeKeywordsEqual` cannot fail and doesn't need error handling. `PyUnicode_Compare` can fail.